File: blender_addon/btp/sdp_envelope.py
"""SDP envelope: BTP1:<gzip+base64url> single-line format.

Compresses a ~1.3 KB SDP into ~800 chars so it fits comfortably in a
single clipboard paste / messenger line. Accepts raw SDP (v=0...) on the
input side so old PWA clients keep working during migration.

Format: `BTP1:` prefix + URL-safe base64 (no padding) of gzipped SDP UTF-8 bytes.
"""
import base64
import gzip


ENVELOPE_PREFIX = "BTP1:"


def encode(sdp: str) -> str:
    gz = gzip.compress(sdp.encode("utf-8"), compresslevel=9)
    b64 = base64.urlsafe_b64encode(gz).decode("ascii").rstrip("=")
    return ENVELOPE_PREFIX + b64


def decode(envelope: str) -> str:
    """Accept either a BTP1 envelope or raw SDP (v=0...). Returns raw SDP."""
    s = envelope.strip()
    if s.startswith("v=0"):
        return s
    if not s.startswith(ENVELOPE_PREFIX):
        raise ValueError(
            f"Unrecognized SDP envelope. Expected '{ENVELOPE_PREFIX}...' "
            f"or raw SDP starting with 'v=0'."
        )
    payload = s[len(ENVELOPE_PREFIX):]
    padded = payload + "=" * (-len(payload) % 4)
    gz = base64.urlsafe_b64decode(padded)
    return gzip.decompress(gz).decode("utf-8")


def is_envelope(s: str) -> bool:
    return s.strip().startswith(ENVELOPE_PREFIX)

File: demo/sdp-envelope.js
// SDP envelope: BTP1:<gzip+base64url> single-line format.
// Mirror of blender_addon/btp/sdp_envelope.py. Keep them in lockstep.

export const ENVELOPE_PREFIX = "BTP1:";

export async function encodeSDP(sdp) {
  const input = new TextEncoder().encode(sdp);
  const cs = new CompressionStream("gzip");
  const writer = cs.writable.getWriter();
  writer.write(input);
  writer.close();
  const gz = new Uint8Array(await new Response(cs.readable).arrayBuffer());
  let b64 = btoa(String.fromCharCode(...gz));
  b64 = b64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
  return ENVELOPE_PREFIX + b64;
}

export async function decodeSDP(envelope) {
  const s = envelope.trim();
  if (s.startsWith("v=0")) return s;
  if (!s.startsWith(ENVELOPE_PREFIX)) {
    throw new Error(`Unrecognized SDP envelope. Expected '${ENVELOPE_PREFIX}...' or raw SDP starting with 'v=0'.`);
  }
  let b64 = s.slice(ENVELOPE_PREFIX.length);
  b64 = b64.replace(/-/g, "+").replace(/_/g, "/");
  while (b64.length % 4) b64 += "=";
  const gz = Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
  const ds = new DecompressionStream("gzip");
  const writer = ds.writable.getWriter();
  writer.write(gz);
  writer.close();
  return new TextDecoder().decode(await new Response(ds.readable).arrayBuffer());
}

export function isEnvelope(s) {
  return s.trim().startsWith(ENVELOPE_PREFIX);
}
